### PR TITLE
Make `resolve` a stack module

### PR DESCRIPTION
**Pull request by olix0r**
The `proxy::http::balance` module uses the `proxy::resolve::Resolve`
trait to implement a `Discover`.

This coupling between the balance and resolve modules prevents
integrating the destination profile API such that there is a per-route,
per-endpoint stack.

This change makes the `balance` stack generic over a stack that produces
a `Discover`. The `resolve` module now implements a stack that produces
a `Discover` and is generic over a per-endpoint stack.

### CODE REVIEW

File: src/app/main.rs [hunk 1]
```diff
@@ -252,31 +254,25 @@ where
...
-                    resolver,
-                )))
-                .and_then(outbound::orig_proto_upgrade::Layer::new())
-                .and_then(outbound::tls_config::Layer::new(tls_client_config))
-                .and_then(proxy::http::metrics::Layer::new(
-                    http_metrics,
-                    classify::Classify,
-                ))
+                .and_then(balance::layer())
+                .and_then(resolve::layer(Resolve::new(resolver)))
```
Comment (hawkw) [inline src/app/main.rs]: Very minor, but: 👍 on changing these to free functions! definitely more readable!

File: src/lib.rs [hunk 2]
```diff
@@ -2,6 +2,9 @@
 #![cfg_attr(feature = "cargo-clippy", allow(new_without_default_derive))]
 #![deny(warnings)]
 
+// Stack type inference requires a deeper recursion limit
+#![recursion_limit="128"]
```
Comment (hawkw) [inline src/lib.rs]: I wonder if breaking up the router stack into a series of separate `let` bindings would help with this?

File: src/proxy/http/balance.rs [hunk 3]
```diff
@@ -2,69 +2,40 @@ extern crate tower_balance;
...
-/// build a service for each endpoint.
-pub struct Service<R: Resolution, M: svc::Stack<R::Endpoint>> {
-    resolution: R,
-    make: M,
 }
 
 // === impl Layer ===
 
-impl<T, R, M, A, B> Layer<T, R, M>
-where
```
Comment (hawkw) [inline src/proxy/http/balance.rs]: awesome to see all these type parameters just go away! :D

File: src/lib.rs (hunk 2 above)
Comment (olix0r) [reply src/lib.rs]: this will be removed in a followup

**Pull request merged**